### ISSUE TITLE
Import from JSON failing

### ISSUE THREAD

**Issue by gauden** [digitalmethodsinitiative/dmi-tcat#88]
I am trying to import a set of JSON objects I collected by other means into DMI-TCAT; the script provided failed. 

As a test, I provided TCAT with a stream of JSON objects created by [`streamR`, the R package](http://cran.r-project.org/web/packages/streamR/streamR.pdf) suggested by the documentation of DMI-TCAT as the source of typical JSON object files for this purpose. I used [a script by the author](http://pablobarbera.com/blog/archives/1.html) of `streamR` to capture the tweets and ensure I had a canonical example of tweets to import. Here are the first few:

```
{"created_at":"Sun Feb 01 22:00:52 +0000 2015","id":562007704424972289,"id_str":"562007704424972289","text":"@SputnikInt Obama is such a puppet.","source":"\u003ca href=\"http:\/\/twitter.com\" rel=\"nofollow\"\u003eTwitter Web Client\u003c\/a\u003e","truncated":false,"in_reply_to_status_id":562007538665664513,"in_reply_to_status_id_str":"562007538665664513","in_reply_to_user_id":34262462,"in_reply_to_user_id_str":"34262462","in_reply_to_screen_name":"SputnikInt","user":{"id":2544278935,"id_str":"2544278935","name":"hilary chekhovian","screen_name":"chekhovian13","location":"","url":null,"description":null,"protected":false,"verified":false,"followers_count":23,"friends_count":128,"listed_count":2,"favourites_count":97,"statuses_count":533,"created_at":"Tue Jun 03 20:37:30 +0000 2014","utc_offset":null,"time_zone":null,"geo_enabled":false,"lang":"en","contributors_enabled":false,"is_translator":false,"profile_background_color":"C0DEED","profile_background_image_url":"http:\/\/abs.twimg.com\/images\/themes\/theme1\/bg.png","profile_background_image_url_https":"https:\/\/abs.twimg.com\/images\/themes\/theme1\/bg.png","profile_background_tile":false,"profile_link_color":"0084B4","profile_sidebar_border_color":"C0DEED","profile_sidebar_fill_color":"DDEEF6","profile_text_color":"333333","profile_use_background_image":true,"profile_image_url":"http:\/\/pbs.twimg.com\/profile_images\/514168251345956864\/P_9w7TBA_normal.jpeg","profile_image_url_https":"https:\/\/pbs.twimg.com\/profile_images\/514168251345956864\/P_9w7TBA_normal.jpeg","default_profile":true,"default_profile_image":false,"following":null,"follow_request_sent":null,"notifications":null},"geo":null,"coordinates":null,"place":null,"contributors":null,"retweet_count":0,"favorite_count":0,"entities":{"hashtags":[],"trends":[],"urls":[],"user_mentions":[{"screen_name":"SputnikInt","name":"Sputnik","id":34262462,"id_str":"34262462","indices":[0,11]}],"symbols":[]},"favorited":false,"retweeted":false,"possibly_sensitive":false,"filter_level":"low","lang":"en","timestamp_ms":"1422828052990"}
{"created_at":"Sun Feb 01 22:00:53 +0000 2015","id":562007704852787202,"id_str":"562007704852787202","text":"This why I rock with Obama... He so Playa y'all just don't realize it https:\/\/t.co\/NiRuSfYiRz","source":"\u003ca href=\"http:\/\/twitter.com\" rel=\"nofollow\"\u003eTwitter Web Client\u003c\/a\u003e","truncated":false,"in_reply_to_status_id":null,"in_reply_to_status_id_str":null,"in_reply_to_user_id":null,"in_reply_to_user_id_str":null,"in_reply_to_screen_name":null,"user":{"id":195912415,"id_str":"195912415","name":"Black Dad Dont Shoot","screen_name":"ImTw_EATin","location":"Searching for Afrika","url":null,"description":"#ProudParent #SeeRed #DaBears #Chicago I represent for the seat that Rosa Parks sat, where Malcolm X was shot, where Martin Luther was popped!!!!","protected":false,"verified":false,"followers_count":1101,"friends_count":1066,"listed_count":4,"favourites_count":7684,"statuses_count":75092,"created_at":"Mon Sep 27 21:50:30 +0000 2010","utc_offset":-21600,"time_zone":"Central Time (US & Canada)","geo_enabled":false,"lang":"en","contributors_enabled":false,"is_translator":false,"profile_background_color":"4A913C","profile_background_image_url":"http:\/\/pbs.twimg.com\/profile_background_images\/545829964193349633\/XYzULf7s.jpeg","profile_background_image_url_https":"https:\/\/pbs.twimg.com\/profile_background_images\/545829964193349633\/XYzULf7s.jpeg","profile_background_tile":true,"profile_link_color":"DD2E44","profile_sidebar_border_color":"000000","profile_sidebar_fill_color":"DDEEF6","profile_text_color":"333333","profile_use_background_image":true,"profile_image_url":"http:\/\/pbs.twimg.com\/profile_images\/554896006798528512\/fL3BGGE2_normal.jpeg","profile_image_url_https":"https:\/\/pbs.twimg.com\/profile_images\/554896006798528512\/fL3BGGE2_normal.jpeg","profile_banner_url":"https:\/\/pbs.twimg.com\/profile_banners\/195912415\/1420946850","default_profile":false,"default_profile_image":false,"following":null,"follow_request_sent":null,"notifications":null},"geo":null,"coordinates":null,"place":null,"contributors":null,"retweet_count":0,"favorite_count":0,"entities":{"hashtags":[],"trends":[],"urls":[{"url":"https:\/\/t.co\/NiRuSfYiRz","expanded_url":"https:\/\/vine.co\/v\/Mh2m9Wl1hEi","display_url":"vine.co\/v\/Mh2m9Wl1hEi","indices":[70,93]}],"user_mentions":[],"symbols":[]},"favorited":false,"retweeted":false,"possibly_sensitive":true,"filter_level":"low","lang":"en","timestamp_ms":"1422828053092"}
{"created_at":"Sun Feb 01 22:00:53 +0000 2015","id":562007705338908672,"id_str":"562007705338908672","text":"RT @OlivierCiappa: Qui vient dans mon Pays la\u00efque voil\u00e9e, me verra dans le sien, comme je suis ts ls jours. La tol\u00e9rance n'est pas \u00e0 sens u\u2026","source":"\u003ca href=\"http:\/\/twitter.com\/download\/android\" rel=\"nofollow\"\u003eTwitter for Android\u003c\/a\u003e","truncated":false,"in_reply_to_status_id":null,"in_reply_to_status_id_str":null,"in_reply_to_user_id":null,"in_reply_to_user_id_str":null,"in_reply_to_screen_name":null,"user":{"id":584682844,"id_str":"584682844","name":"S\u00e9bastien Lbrt","screen_name":"Seblbrt","location":"Co\u00ebx, Vend\u00e9e, France","url":null,"description":null,"protected":false,"verified":false,"followers_count":123,"friends_count":788,"listed_count":2,"favourites_count":151,"statuses_count":1155,"created_at":"Sat May 19 11:42:37 +0000 2012","utc_offset":null,"time_zone":null,"geo_enabled":true,"lang":"fr","contributors_enabled":false,"is_translator":false,"profile_background_color":"C0DEED","profile_background_image_url":"http:\/\/abs.twimg.com\/images\/themes\/theme1\/bg.png","profile_background_image_url_https":"https:\/\/abs.twimg.com\/images\/themes\/theme1\/bg.png","profile_background_tile":false,"profile_link_color":"0084B4","profile_sidebar_border_color":"C0DEED","profile_sidebar_fill_color":"DDEEF6","profile_text_color":"333333","profile_use_background_image":true,"profile_image_url":"http:\/\/pbs.twimg.com\/profile_images\/2619994500\/image_normal.jpg","profile_image_url_https":"https:\/\/pbs.twimg.com\/profile_images\/2619994500\/image_normal.jpg","default_profile":true,"default_profile_image":false,"following":null,"follow_request_sent":null,"notifications":null},"geo":null,"coordinates":null,"place":null,"contributors":null,"retweeted_status":{"created_at":"Sun Feb 01 20:00:45 +0000 2015","id":561977474087919616,"id_str":"561977474087919616","text":"Qui vient dans mon Pays la\u00efque voil\u00e9e, me verra dans le sien, comme je suis ts ls jours. La tol\u00e9rance n'est pas \u00e0 sens unique\"Michelle Obama","source":"\u003ca href=\"http:\/\/twitter.com\" rel=\"nofollow\"\u003eTwitter Web Client\u003c\/a\u003e","truncated":false,"in_reply_to_status_id":null,"in_reply_to_status_id_str":null,"in_reply_to_user_id":null,"in_reply_to_user_id_str":null,"in_reply_to_screen_name":null,"user":{"id":1012155314,"id_str":"1012155314","name":"Olivier Ciappa","screen_name":"OlivierCiappa","location":"Paris","url":"http:\/\/www.olivierciappa.com","description":"R\u00e9alisateur - photographe - cr\u00e9ateur de timbres de collection (dont la nouvelle Marianne) et parfois un peu activiste engag\u00e9...","protected":false,"verified":false,"followers_count":5947,"friends_count":308,"listed_count":52,"favourites_count":83,"statuses_count":2173,"created_at":"Fri Dec 14 23:15:26 +0000 2012","utc_offset":3600,"time_zone":"Paris","geo_enabled":true,"lang":"fr","contributors_enabled":false,"is_translator":false,"profile_background_color":"7B7361","profile_background_image_url":"http:\/\/pbs.twimg.com\/profile_background_images\/749773786\/7f4943fa66bfc6ac500719905b2ea146.jpeg","profile_background_image_url_https":"https:\/\/pbs.twimg.com\/profile_background_images\/749773786\/7f4943fa66bfc6ac500719905b2ea146.jpeg","profile_background_tile":true,"profile_link_color":"C2BAA8","profile_sidebar_border_color":"FFFFFF","profile_sidebar_fill_color":"DDEEF6","profile_text_color":"333333","profile_use_background_image":true,"profile_image_url":"http:\/\/pbs.twimg.com\/profile_images\/554790361332060160\/lWrbNS6r_normal.jpeg","profile_image_url_https":"https:\/\/pbs.twimg.com\/profile_images\/554790361332060160\/lWrbNS6r_normal.jpeg","profile_banner_url":"https:\/\/pbs.twimg.com\/profile_banners\/1012155314\/1418215524","default_profile":false,"default_profile_image":false,"following":null,"follow_request_sent":null,"notifications":null},"geo":null,"coordinates":null,"place":null,"contributors":null,"retweet_count":97,"favorite_count":44,"entities":{"hashtags":[],"trends":[],"urls":[],"user_mentions":[],"symbols":[]},"favorited":false,"retweeted":false,"possibly_sensitive":false,"filter_level":"low","lang":"fr"},"retweet_count":0,"favorite_count":0,"entities":{"hashtags":[],"trends":[],"urls":[],"user_mentions":[{"screen_name":"OlivierCiappa","name":"Olivier Ciappa","id":1012155314,"id_str":"1012155314","indices":[3,17]}],"symbols":[]},"favorited":false,"retweeted":false,"possibly_sensitive":false,"filter_level":"low","lang":"fr","timestamp_ms":"1422828053208"}
```

For each tweet I am trying to import the `dmi-tcat/capture/common/functions.php` script generates a series of errors and warnings:

```
PHP Notice:  Trying to get property of non-object in /Applications/XAMPP/xamppfiles/htdocs/dmi-tcat/capture/common/functions.php on line 1264
PHP Notice:  Trying to get property of non-object in /Applications/XAMPP/xamppfiles/htdocs/dmi-tcat/capture/common/functions.php on line 1265
PHP Notice:  Trying to get property of non-object in /Applications/XAMPP/xamppfiles/htdocs/dmi-tcat/capture/common/functions.php on line 1266
PHP Notice:  Trying to get property of non-object in /Applications/XAMPP/xamppfiles/htdocs/dmi-tcat/capture/common/functions.php on line 1267
PHP Notice:  Trying to get property of non-object in /Applications/XAMPP/xamppfiles/htdocs/dmi-tcat/capture/common/functions.php on line 1268
PHP Warning:  Attempt to modify property of non-object in /Applications/XAMPP/xamppfiles/htdocs/dmi-tcat/capture/common/functions.php on line 1304
PHP Notice:  Trying to get property of non-object in /Applications/XAMPP/xamppfiles/htdocs/dmi-tcat/capture/common/functions.php on line 1309
PHP Notice:  Trying to get property of non-object in /Applications/XAMPP/xamppfiles/htdocs/dmi-tcat/capture/common/functions.php on line 1320
PHP Notice:  Trying to get property of non-object in /Applications/XAMPP/xamppfiles/htdocs/dmi-tcat/capture/common/functions.php on line 1331
```

For ease of reference, the relevant lines are:

- lines 1264-8 (`PHP Notice:  Trying to get property of non-object`):

```{php}
'hashtags' => count($obj['tweet']->hashtags),
'urls' => count($obj['tweet']->urls),
'mentions' => count($obj['tweet']->user_mentions),
'withheld' => count($obj['tweet']->withheld_in_countries),
'places' => count($obj['tweet']->places)
```

- line 1304 (`PHP Warning:  Attempt to modify property of non-object`):

```{php}
$tweetq->bindParam($tweeti++, $t->$f);
```

- lines 1309, 1320, 1331 (`PHP Notice:  Trying to get property of non-object`):

And the other three errors are generated when the tweet object's fields are accessed (on line 1309 `$t->hashtags`, on line 1320  `$t->user_mentions`, and on line 1331 `$t->urls`). 

The whole thing is running on a Mac OSX system (Yosemite) on a XAMPP stack with PHP v5.6.3. Analysis and capture work perfectly. It just seems that captured tweets enter the system smoothly but imported tweets are somehow converted to arrays not objects?

**Comment by gauden** [comment]
As additional information, the import script would not even start to run unless I imported the tweets as arrays by editing line 62 of `dmi-tcat/import/import-jsondump.php` to this:

```{php}
$tweet = json_decode($buffer, $assoc=true);
```

Without the `$assoc` flag, import stops immediately with a fatal error:

```
PHP Fatal error:  Cannot use object of type stdClass as array in 
/Applications/XAMPP/xamppfiles/htdocs/dmi-tcat/capture/common/functions.php on line 928
```

**Comment by dentoir** [comment]
This should be fixed in latest commit: 0a491830cd

**Comment by gauden** [comment]
That works. The issue is closed. Thank you.
